alternator/streams.cc

```diff
@@ -145,19 +145,24 @@ future<alternator::executor::request_return_type> alternator::executor::list_str
     auto table = find_table(_proxy, request);
     auto db = _proxy.data_dictionary();
     auto cfs = db.get_tables();
-    auto i = cfs.begin();
-    auto e = cfs.end();
 
     if (limit < 1) {
         throw api_error::validation("Limit must be 1 or more");
     }
 
-    // TODO: the unordered_map here is not really well suited for partial
-    // querying - we're sorting on local hash order, and creating a table
-    // between queries may or may not miss info. But that should be rare,
-    // and we can probably expect this to be a single call.
+    // # 12601 (maybe?) - sort the set of tables on ID. This should ensure we never
+    // generate duplicates in a paged listing here. Can obviously miss things if they 
+    // are added between paged calls and end up with a "smaller" UUID/ARN, but that 
+    // is to be expected.
+    std::sort(cfs.begin(), cfs.end(), [](const data_dictionary::table& t1, const data_dictionary::table& t2) {
+        return t1.schema()->id().uuid() < t2.schema()->id().uuid();
+    });
+
```

Comment:
The whole thing (even before the patch) has quadratic complexity, now it's O(n^2 log n). The number of pages is O(n), copying the tables from data_dictionary is O(n), and now this sort is O(n log n).

It doesn't matter if the number of pages is always very low. Is this the case? A different way to phrase the question: how many tables can a page contain?

If it's a problem, we'll need to define an iterative interface to data_dictionary that has a consistent sort order.

Reply:
‎The "typical" case is to have an unlimited or very large page size and a "reasonably small" number of tables (Scylla doesn't support a huge number of tables anyway), and call ListStreams only rarely if at all, so the performance of this doesn't matter. But I agree that in the worst case doing ListStreams in tiny pages now needs O(n^2 log n), which isn't much worse than O(n^2). You're right that it would be nice for db::get_tables(), which already returns a vector with some order - to say something about this order. Or create a new function that returns the list in a consistent order. The question is how to do this efficiently (maybe replace the unsorted map of tables by a sorted map?).

@elcallio looking at the code again there is something that concerns me even more. I assume that almost always, ListStreams will be called for a specific table. Ideally in this case, it should be O(1), not O(n log n) . But currently it isn't - we iterate over all tables (after sorting them). Couldn't we find a way to make it O(1)? I think this case will be much more common then listing all the streams of all the tables in the system.

Reply:
If we only query one table, even the copy of "tables" seems excessive. Technically, we could just check if a CDC table exists for the given table, and just return it. 

Indeed the default paging is in fact unlimited here. I wonder if this is ok, since we are supposed to be limited in how much payload we can generate in a single call, no? But again, not exactly heavy on the output this method. 

I am not sure if this method is such a hotspot though, so not eager to optimize it too much until shown its needed. Again, iterating ~100 tables now and then is just good for the cpu (get some air). :-)

Reply:
Maybe all we need is some yields to avoid stalls (but need to exercise care in case the schema is updated while we yield).